managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
+
+A unified interface needs to be capable of supporting both of these kinds of memory architectures in such a way that any requirements for cache coherancy is abstracted by the interface and associated coherency model.
+
+### 2.2. Must support different levels of synchronisation
+
+Different memory architectures, both unified and discrete; may support granularity at different levels. This can either be at the course grained level; where synchronisation between memory regions is performed at the execution function level, or this can be at the fine grained level; where synchronisation between memory regions is performed at the instruction level, allowing access to memory from different devices concurrently, using atomics to specify read and write order.
+
+A unified interface needs to be capable of supporting both these levels of synchronisation.
 
-## Proposed Additions
+### 2.3. Must support different data flow models
 
-### Requirements on Execution Context
+There are also many different data flow models that can be used to describe data and data dependenies in relation to functions of execution. Though it would be impossible to cover all possible models, it's important to make a unified inteface as flexible and utilisable as possible, therfore facilitating a wide range of data flow models.
 
-For the purposes of this paper, it is necessary to introduce requirements on the **execution context**; the object associated with an **executor** which encapsulates the underlying execution resource on which functions are executed on.
+One such data flow model that C\+\+ is moving towards is that of futures and continuations; the idiom that the invocation of an asycnhornous task returns a future object that represents a dependency on the completion of that task that can be waiting or used to create a continuation task. This allows tasks to be linked together by their dependencies on each other.
 
-The **execution context** must encapsulate both the execution agents and a global memory region that is accessible from those execution agents. It is required that the memory region of the **execution context** be cache coherent across the execution agents created from the same invocation of an execution function. It is not required that the memory region of the **execution context** be cache coherent across execution agents created from different execution functions or on different **execution contexts**.
+Another data flow model is one that uses a runtime to create a graph of tasks, which can be optimised.
 
-*Note: some systems are capable of providing cache coherency across execution functions or **execution contexts** through the availability of shared virtual memory addressing or shared physical memory. This feature is discussed in more detail in the future work section.*
+Another data flow model is one which simply submits work to be executed and waits for it all to complete, typically called fire and forget execution.
```

Comment:
This was more of a special case of another more general data-flow model, it was not a good example so I have removed it.